### PR TITLE
Missing metadata

### LINKED ISSUE #1119: Sending a raw message, that fails, results in SC not being able to return the failed message to SP
So here https://github.com/Particular/ServiceControl/blob/master/src/ServiceControl/MessageFailures/Api/GetErrorById.cs#L55

It is assumed that many headers exist and no defensive code is done

When sending a raw message using this approach https://docs.particular.net/transports/sql/operations-scripting#native-send-the-native-send-helper-methods-in-c

The only required header for NSB endpoints to successfully process a message is `NServiceBus.EnclosedMessageTypes`. But, if the message processing fails, this results in `processingAttempt.MessageMetadata` in SC missing the `SendingEndpoint` and `TimeSent` headers.

When SP queries for this failed message SC throws a `KeyNotFoundException`

```
System.Collections.Generic.KeyNotFoundException: The given key was not present in the dictionary.
   at System.ThrowHelper.ThrowKeyNotFoundException()
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at ServiceControl.MessageFailures.Api.GetErrorById.Map(FailedMessage message, IAsyncDocumentSession session) in C:\BuildAgent\work\7189a56f9f44affc\src\ServiceControl\MessageFailures\Api\GetErrorById.cs:line 55
   at ServiceControl.MessageFailures.Api.GetErrorById.<<-ctor>b__0_1>d.MoveNext() in C:\BuildAgent\work\7189a56f9f44affc\src\ServiceControl\MessageFailures\Api\GetErrorById.cs:line 44
```

### PR DESCRIPTION
**Pull request by mikeminutillo**
Fixes #1119 and #1120

### CODE REVIEW

File: src/ServiceControl/Infrastructure/MetadataExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
+﻿namespace ServiceControl
+{
+    using System;
+    using System.Collections.Generic;
+
+    static class MetadataExtensions
+    {
+        public static T MaybeGet<T>(this IDictionary<string, object> metadata, string key)
```
Comment (SzymonPobiega) [inline src/ServiceControl/Infrastructure/MetadataExtensions.cs]: How about `GetOrDefault` ?

Comment (danielmarbach) [reply src/ServiceControl/Infrastructure/MetadataExtensions.cs]: I would prefer GetOrDefault as well

File: src/ServiceControl/Infrastructure/MetadataExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+            {
+                return foundValue?.ToString();
+            }
+
+            return null;
+        }
+
+        public static DateTime? AsMaybeDateTime(this string s)
+        {
+            if (s == null) return null;
```
Comment (danielmarbach) [inline src/ServiceControl/Infrastructure/MetadataExtensions.cs]: Should this have braces?

File: src/ServiceControl/Infrastructure/MetadataExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+            DateTime dt;
+            if (DateTime.TryParse(s, out dt))
```
Comment (danielmarbach) [inline src/ServiceControl/Infrastructure/MetadataExtensions.cs]: Wouldn't we want to specify the datetime format we want to see here as well as the datetime style and the invariant culture?

File: src/ServiceControl/Infrastructure/MetadataExtensions.cs (hunk 1 above)
Comment (mikeminutillo) [reply src/ServiceControl/Infrastructure/MetadataExtensions.cs]: What about the `MaybeGetAsString()`? `GetOrNullAsString()`?

File: src/ServiceControl/Infrastructure/MetadataExtensions.cs (hunk 3 above)
Comment (mikeminutillo) [reply src/ServiceControl/Infrastructure/MetadataExtensions.cs]: Possibly.

The previous code was just `DateTime.Parse` but it was immediately after a `ToString()`. We might be able to just turn it into a `GetOrDefault<DateTime>()`. I'm not sure why it was being round-tripped through `System.String`.

File: src/ServiceControl/Infrastructure/MetadataExtensions.cs (hunk 1 above)
Comment (WilliamBZA) [reply src/ServiceControl/Infrastructure/MetadataExtensions.cs]: `GetAsStringOrNull`

File: src/ServiceControl/Infrastructure/MetadataExtensions.cs (hunk 3 above)
Comment (danielmarbach) [reply src/ServiceControl/Infrastructure/MetadataExtensions.cs]: Maybe then doing TryParse should work. I have not enough knowledge here to say it should work or it won't

**Comment by danielmarbach** [comment]
@mikeminutillo I believe this should also fix #1120 added it to the description

**Comment by SimonCropp** [comment]
@mikeminutillo @danielmarbach thanks guys

**Pull request merged**